### DISCUSSION TITLE
【API设计】应该如何加入“禁止宽屏重定位“的功能

### DISCUSSION THREAD

**Discussion by timongh** [the1812/Bilibili-Evolved#2784] [综合讨论 / General]
#### 概述

最近找到了实现“禁止宽屏重定位” 的方式，然而该功能可能有多个组件需要（见 #483）。我没有能力设计这个 api，因此发起讨论，同时也是寻找有能力的人接手。

#### 详细内容

该功能的实现需要重新包装一个全局函数，而组件中，有的需要永久开启，有的只要临时开启，有的需要该功能，有的可能不需要。

因此，要么设计一个统一的 api 进行管理，要么写好文档让用户自行管理。

我的倾向是，不管是不是让用户自行管理，有一个统一 api 都是好的，这样的代码比较“干净”。但是，我思考了一段时间，发现自己无法设计好这个 api。因此希望大家能各抒己见（如果能直接搞定的话，那就再好不过了!!）。

该功能的实现方式如下：

```typescript
const orgPlayerWideWin = unsafeWindow.player_widewin

unsafeWindow.player_widewin = () => {
  // B 站宽屏函数内部会调用 window.scrollTo() 来进行重定位。要禁止重定位行为就需要抑制该函数
  // 原理：通过 height 和 overflow 两个属性，让 body 正好填满视口而不溢出。
  // 此时，再对 window 滚动是没有意义的，因此 window.scrollTo 函数失效，重定位也失效。
  // 而 body 高度被限制后，其顶部内容也会回到视口的顶部，看起来就像内容滚回了头部；
  // 因此，为了不跳回顶部，必须提前记录滚动位置，并在每次样式改变时恢复滚动数据。
  const theScroll = window.scrollY

  const newStyle = { height: '100vh', overflow: 'scroll' }
  const orgStyle = lodash.pick(document.body.style, Object.keys(newStyle))

  lodash.assign(document.body.style, newStyle)
  document.body.scrollTo(0, theScroll)

  orgPlayerWideWin()

  lodash.assign(document.body.style, orgStyle)
  window.scrollTo(0, theScroll)
}
```

**Comment by the1812** [comment]
我倒是想试试临时 Hook 下 `window.scrollTo`, 另外问下有哪些功能需要永久开启禁止宽屏重定位?

**Reply by timongh** [reply]
开启“播放器置顶”后，再切换宽屏会有奇怪的定位行为；而如果全局禁止掉重定位就没问题了。

另外该功能本身也可以单独作为一个组件使用。

而“默认播放器模式”是否禁止重定位，要看用户自己的偏好。

这几个组件的需求各不相同，如果没有统一的 API，可能有冲突和混乱。比如，两个组件同时禁止重定位，那么 `window.player_widewin` 就会被套上两层。又或者，用户只希望在“默认播放器模式”临时开启重定位，其他时候都禁止掉。

当然，也有可能是我想太多。说不定直接在文档里写清楚，让用户自行决定就已经是足够优秀的解决方案了呢。

**Reply by timongh** [reply]
对了，`window.player_widewin` 很好定位。打开视频页面，在开发人员工具中选中中断于 `scroll` 事件，此时点击宽屏按钮就会立刻中断。在调用堆栈中可以看到 `player_widewin` 和 `window.player_widewin` 两个函数。后者是我们要 hook 的函数；前者是后者的底层函数，其内部就有对 `window.scrollTo` 的调用。

**Comment by the1812** [comment]
久等了, Hook 确实可以用, 支持暂时或永久.
https://github.com/the1812/Bilibili-Evolved/blob/4009bf9d54f9bf26c5d44b32b7f11b1880b2dce7/src/core/utils/index.ts#L419-L436
https://github.com/the1812/Bilibili-Evolved/blob/4009bf9d54f9bf26c5d44b32b7f11b1880b2dce7/registry/lib/components/video/player/default-mode/index.ts#L20-L22

**Reply by timongh** [reply]
原来能够直接改 window 下的函数啊。我当初在控制台试过，没成功，应该是当时我哪里写错了吧。所以才弄了上面那么个别扭的方式。

**Comment by CKylinMC** [comment]
我觉得其实也不一定非要hook那个player_widewin，万一B站改个什么逻辑又坏了。
如果只是针对player_widewin函数做屏蔽，或许也可以检查`(new Error).stack`里是不是包含player_widewin，不修改B站的方法而只hook那个scrollTo。

**Reply by the1812** [reply]
现在就是采用 hook scrollTo 的方案
